ProcessLocksScreen.c

```diff
@@ -0,0 +1,180 @@
+}
+
+void ProcessLocksScreen_delete(Object* this) {
+   free(InfoScreen_done((InfoScreen*)this));
+}
+
+static void ProcessLocksScreen_draw(InfoScreen* this) {
+   InfoScreen_drawTitled(this, "Snapshot of file locks of process %d - %s", ((ProcessLocksScreen*)this)->pid, this->process->comm);
+}
+
+static FileLocks_ProcessData* ProcessLocksScreen_getProcessData(pid_t pid) {
+   FileLocks_ProcessData* pdata = xCalloc(1, sizeof(FileLocks_ProcessData));
+
+   FILE* f = fopen(PROCDIR "/locks", "r");
+   if (!f) {
+      pdata->error = 127;
+      return pdata;
+   }
+
+   char buffer[1024];
+   FileLocks_LockData** data_ref = &pdata->locks;
+   while(fgets(buffer, sizeof(buffer), f)) {
+      if (!strchr(buffer, '\n')) {
+         continue;
+      }
+
+      int lock_id;
+      char lock_type[16];
+      char lock_excl[16];
+      char lock_rw[16];
+      pid_t lock_pid;
+      unsigned int lock_inode[3];
+      char lock_start[25];
+      char lock_end[25];
+
+      if (10 != sscanf(buffer, "%d:  %15s  %15s %15s %d %x:%x:%u %24s %24s",
+         &lock_id, lock_type, lock_excl, lock_rw, &lock_pid,
+         &lock_inode[0], &lock_inode[1], &lock_inode[2],
+         lock_start, lock_end)
+      ) {
+         continue;
+      }
+
+      if (pid != lock_pid) {
+         continue;
+      }
+      FileLocks_LockData* ldata = xCalloc(1, sizeof(FileLocks_LockData));
+      FileLocks_Data* data = &ldata->data;
+      data->id = lock_id;
+      data->data[0] = xStrdup(lock_type);
+      data->data[1] = xStrdup(lock_excl);
+      data->data[2] = xStrdup(lock_rw);
+      data->inode[0] = lock_inode[0];
+      data->inode[1] = lock_inode[1];
+      data->inode[2] = lock_inode[2];
+      data->start = strtoull(lock_start, NULL, 10);
+      if (0 != strcmp(lock_end, "EOF")) {
+         data->end = strtoull(lock_end, NULL, 10);
+      } else {
+         data->end = ULLONG_MAX;
+      }
+
+      *data_ref = ldata;
+      data_ref = &ldata->next;
+   }
+
+   fclose(f);
+   return pdata;
+}
+
+static inline void FileLocks_Data_clear(FileLocks_Data* data) {
+   for (size_t i = 0; i < ARRAYSIZE(data->data); i++)
+      free(data->data[i]);
+}
+
+static void ProcessLocksScreen_scan(InfoScreen* this) {
+   Panel* panel = this->display;
+   int idx = Panel_getSelectedIndex(panel);
+   Panel_prune(panel);
+   FileLocks_ProcessData* pdata = ProcessLocksScreen_getProcessData(((ProcessLocksScreen*)this)->pid);
+   if (pdata->error == 127) {
+      InfoScreen_addLine(this, "Could not read file locks.");
+   } else {
+      FileLocks_LockData* ldata = pdata->locks;
+      while (ldata) {
+         FileLocks_Data* data = &ldata->data;
+
+         char entry[128];
+         if (ULLONG_MAX == data->end) {
+            xSnprintf(entry, sizeof(entry), "%10d: %-10s %-10s %-10s %02x:%02x:%010u %20"PRIu64" %20s",
```

Comment:
`"%10d: %-10s %-10s %-10s %02x:%02x:%010u %20"PRIu64" %20s"`
this`..^..` colon looks weird to me; maybe just two spaces between ID and type?